src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -175,6 +181,301 @@ static inline ASR::asr_t* create_UnaryFunction(Allocator& al, const Location& lo
+                        ASR::cmpopType::Eq,
+                        EXPR(ASR::make_StringItem_t(al, loc, args[1],
+                            EXPR(ASR::make_IntegerBinOp_t(al, loc, pi_len,
+                                ASR::binopType::Add,
+                                EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                            int_type, nullptr)),
+                        char_type, nullptr)),
+                    logical_type, nullptr));
+                Vec<ASR::stmt_t *> if_body_1; if_body_1.reserve(al, 1);
+                if_body_1.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                    pi_len, EXPR(ASR::make_IntegerBinOp_t(al, loc, pi_len,
+                    ASR::binopType::Add, EXPR(ASR::make_IntegerConstant_t(al,
+                    loc, 1, int_type)), int_type, nullptr)), nullptr)));
+                if_body_1.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                    EXPR(ASR::make_ListItem_t(al, loc, lps, i, int_type, nullptr)),
+                    pi_len, nullptr)));
+                if_body_1.push_back(al, STMT(ASR::make_Assignment_t(al, loc, i,
+                    EXPR(ASR::make_IntegerBinOp_t(al, loc, i, ASR::binopType::Add,
+                    EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                    int_type, nullptr)), nullptr)));
+                Vec<ASR::stmt_t *> else_body_1; else_body_1.reserve(al, 1);
+                {
+                    Vec<ASR::stmt_t *> if_body_2; if_body_2.reserve(al, 1);
+                    Vec<ASR::stmt_t *> else_body_2; else_body_2.reserve(al, 1);
+                    ASR::expr_t *a_test = EXPR(ASR::make_IntegerCompare_t(al, loc,
+                        pi_len, ASR::cmpopType::NotEq,
+                        EXPR(ASR::make_IntegerConstant_t(al, loc, 0, int_type)),
+                        logical_type, nullptr));
+                    if_body_2.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                        pi_len, EXPR(ASR::make_ListItem_t(al, loc, lps,
+                            EXPR(ASR::make_IntegerBinOp_t(al, loc, pi_len,
+                                ASR::binopType::Sub,
+                                EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                                int_type, nullptr)),
+                            int_type, nullptr)), nullptr)));
+                    else_body_2.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                        i, EXPR(ASR::make_IntegerBinOp_t(al, loc, i,
+                            ASR::binopType::Add,
+                            EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                            int_type, nullptr)), nullptr)));
+                    else_body_1.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
+                        if_body_2.p, if_body_2.n, else_body_2.p, else_body_2.n)));
+                }
+                loop_body.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
```

Comment:
This loop creation can be abstracted into an API just like (just overload `create_do_loop` as you like) 

https://github.com/lfortran/lfortran/blob/6eeb126cde6dcc4a1e73371e0fa3c7a010324b7f/src/libasr/pass/intrinsic_function_registry.h#L55-L116